src/Microsoft.ML.TimeSeries/SequentialAnomalyDetectionTransformBase.cs

```diff
@@ -557,102 +538,89 @@ protected override sealed void InitializeStateCore()
-            /// <param name="length"></param>
-            public SchemaImpl(ISchema schema, int col, string[] names, int length)
+            public Mapper(IHostEnvironment env, SequentialAnomalyDetectionTransformBase<TInput, TState> parent, ISchema inputSchema)
             {
-                Contracts.Assert(length > 0);
-                Contracts.Assert(Utils.Size(names) >= length);
-                Contracts.AssertValue(schema);
-                Contracts.Assert(0 <= col && col < schema.ColumnCount);
-                _parent = schema;
-                _col = col;
-
-                _names = names;
-                _namesLength = length;
-
-                _type = new VectorType(TextType.Instance, _namesLength);
-                Contracts.AssertValue(_type);
-                _getter = GetSlotNames;
-
-                AsSchema = Schema.Create(this);
+                Contracts.CheckValue(env, nameof(env));
+                _host = env.Register(nameof(Mapper));
+                _host.CheckValue(inputSchema, nameof(inputSchema));
+                _host.CheckValue(parent, nameof(parent));
+
+                if (!inputSchema.TryGetColumnIndex(parent.InputColumnName, out var col))
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", parent.InputColumnName);
+
+                var colType = inputSchema.GetColumnType(col);
+                if (colType != NumberType.R4)
+                    throw _host.ExceptSchemaMismatch(nameof(inputSchema), "input", parent.InputColumnName, NumberType.R4.ToString(), colType.ToString());
+
+                _parent = parent;
+                _parentSchema = inputSchema;
+                _slotNames = new VBuffer<ReadOnlyMemory<char>>(4, new[] { "Alert".AsMemory(), "Raw Score".AsMemory(),
+                    "P-Value Score".AsMemory(), "Martingale Score".AsMemory() });
             }
 
-            public bool TryGetColumnIndex(string name, out int col)
+            public Schema.Column[] GetOutputColumns()
             {
-                return _parent.TryGetColumnIndex(name, out col);
+                var meta = new Schema.Metadata.Builder();
+                meta.AddSlotNames(_parent._outputLength, GetSlotNames);
+                var info = new Schema.Column[1];
+                info[0] = new Schema.Column(_parent.OutputColumnName, new VectorType(NumberType.R8, _parent._outputLength), meta.GetMetadata());
+                return info;
             }
 
-            public string GetColumnName(int col)
+            public void GetSlotNames(ref VBuffer<ReadOnlyMemory<char>> dst)
             {
-                return _parent.GetColumnName(col);
+                _slotNames.CopyTo(ref dst, 0, _parent._outputLength);
             }
 
-            public ColumnType GetColumnType(int col)
+            public Func<int, bool> GetDependencies(Func<int, bool> activeOutput)
             {
-                return _parent.GetColumnType(col);
+                return col => activeOutput(0);
```

Comment:
>col => activeOutput(0) [](start = 23, length = 22)

that's incorrect. If `activeOutput(0)`, then you only need one column, not all columns from the input. Let's have a restrictive enough predicate

Reply:
in ctor, save `inputColumnIndex`. Then here you can do `if (activeOutput(0)) return col=>col==_inputColumnIndex; else return col=>false;`

---
In reply to: [227616090](https://github.com/dotnet/machinelearning/pull/1254#discussion_r227616090) [](ancestors = 227616090)

Reply:
thx

---
In reply to: [229022187](https://github.com/dotnet/machinelearning/pull/1254#discussion_r229022187) [](ancestors = 229022187,227616090)